### PR TITLE
DAT-3: BE - Registration API

### PR DESCRIPTION
**Pull request by thanhphuchuynh**
code user registration APIs (response token user) 

image

image

### CODE REVIEW

**Comment by paphuc** [comment]
Please help to use command: `go fmt ...` to format the project

File: internal/app/api/handler/user/user.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
+package userhandler
+
+import (
+	"context"
+	"dating/internal/app/types"
+	"dating/internal/pkg/glog"
+	"dating/internal/pkg/respond"
```
Comment (paphuc) [inline internal/app/api/handler/user/user.go]: Move all datting packages to the bottom

File: internal/app/types/claim.go [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package types
```
Comment (paphuc) [inline internal/app/types/claim.go]: Move types package into api folder

File: internal/app/user/mongodb_repo.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
+package user
```
Comment (paphuc) [inline internal/app/user/mongodb_repo.go]: Create a folder `repositories` in api folder, then move all the mongo_repo.go into it. Then rename this file to user_repo.go

File: internal/app/user/user.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
+package user
```
Comment (paphuc) [inline internal/app/user/user.go]: Please help to create a folder `services `in api folder, then move this package into it

File: internal/app/user/mongodb_repo.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+	session *mgo.Session
+}
+
+func NewMongoRepository(s *mgo.Session) *MongoRepository {
+	return &MongoRepository{
+		session: s,
+	}
+}
+
+// Sign up method
```
Comment (paphuc) [inline internal/app/user/mongodb_repo.go]: Please make a doc  more useful. Ex: This method helps to register new member ...

File: internal/app/user/mongodb_repo.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+		return nil, errors.Wrap(err, "can't insert user")
+	}
+
+	return &types.UserResponseSignUp{
+		Name:  UserSignUp.Name,
+		Email: UserSignUp.Email,
+		Token: tokenString}, nil
+}
+
+func (r *MongoRepository) Login(ctx context.Context, UserLogin types.UserLogin) (*types.UserResponseSignUp, error) {
```
Comment (paphuc) [inline internal/app/user/mongodb_repo.go]: Please add a doc

File: internal/app/user/mongodb_repo.go [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+	s := r.session.Clone()
+	defer s.Close()
+	var User types.User
```
Comment (paphuc) [inline internal/app/user/mongodb_repo.go]: No need to use the public variable here

File: internal/app/user/mongodb_repo.go [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+	if err := r.collection(s).Find(bson.M{"email": UserLogin.Email}).One(&User); err != nil {
+		return nil, errors.Wrap(errors.New("not found email exits"), "email not exists, can't find user")
+	}
+	// isCorrectPassword true
+	isCorrectPassword := isCorrectPassword(UserLogin.Password, User.Password)
+
+	if !isCorrectPassword {
```
Comment (paphuc) [inline internal/app/user/mongodb_repo.go]: No need to create a variable which is used one time. 
It should be: if isCorrectPassword(UserLogin.Password, User.Password)

File: internal/app/user/mongodb_repo.go [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+		Token: tokenString}, nil
+
+}
+
+func (r *MongoRepository) collection(s *mgo.Session) *mgo.Collection {
+	return s.DB("").C("users")
+}
+
+//Generate token
+func GenToken(user types.UserSignUp) (string, error) {
```
Comment (paphuc) [inline internal/app/user/mongodb_repo.go]: In repository, we should handle interaction with DB only. So please help to move the logic into the service. 
About the JWT, it should be in the pkg folder

File: internal/app/member/mongodb_repo.go [hunk 10]
```diff
@@ -30,6 +30,7 @@ func (r *MongoRepository) FindByID(ctx context.Context, id string) (*types.Membe
 	if err := r.collection(s).Find(bson.M{"id": id}).One(&member); err != nil {
 		return nil, errors.Wrap(err, "failed to find the given member from database")
 	}
+
```
Comment (paphuc) [inline internal/app/member/mongodb_repo.go]: Please help to remove the member package if it's not used anymore

File: internal/app/types/user.go [hunk 11]
```diff
@@ -0,0 +1,31 @@
+package types
+
+import "github.com/globalsign/mgo/bson"
+
+type User struct {
+	ID       bson.ObjectId   `json:"_id" bson:"_id,omitempty"`
```
Comment (paphuc) [inline internal/app/types/user.go]: please add more fields: created_at, updated_at into it to keep track history and other tables should too

File: internal/app/types/user.go [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+	Name     string          `json:"name" bson:"name"`
+	Email    string          `json:"email" bson:"email"`
+	Password string          `json:"password" bson:"password"`
+	Avatars  []string        `json:"avatar" bson:"avatars"` // arr path image
+	Gender   string          `json:"gender" bson:"gender"`
+	Country  string          `json:"country" bson:"country"`
+	Hobby    []Hobby         `json:"hobby" bson:"hobby"`
+	About    string          `json:"about" bson:"about"`
+	Like_id  []bson.ObjectId `json:"like_id" bson:"like_id"`
+	Math     []bson.ObjectId `json:"match" bson:"math"`
```
Comment (paphuc) [inline internal/app/types/user.go]: What is this column for?

File: internal/app/types/user.go [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+	ID       bson.ObjectId   `json:"_id" bson:"_id,omitempty"`
+	Name     string          `json:"name" bson:"name"`
+	Email    string          `json:"email" bson:"email"`
+	Password string          `json:"password" bson:"password"`
+	Avatars  []string        `json:"avatar" bson:"avatars"` // arr path image
+	Gender   string          `json:"gender" bson:"gender"`
+	Country  string          `json:"country" bson:"country"`
+	Hobby    []Hobby         `json:"hobby" bson:"hobby"`
+	About    string          `json:"about" bson:"about"`
+	Like_id  []bson.ObjectId `json:"like_id" bson:"like_id"`
```
Comment (paphuc) [inline internal/app/types/user.go]: It should be LikeID

File: internal/app/user/mongodb_repo.go [hunk 14]
```diff
@@ -0,0 +1,120 @@
+package user
+
+import (
+	"context"
+	"dating/internal/app/types"
+	"encoding/base64"
+	"fmt"
+
```
Comment (paphuc) [inline internal/app/user/mongodb_repo.go]: Remove the redundant lines

**Review by paphuc — CHANGES_REQUESTED** [review]

File: internal/app/types/user.go (hunk 12 above)
Comment (thanhphuchuynh) [reply internal/app/types/user.go]: this column for save id users matched. Should change the name to MathID ?.

**Review by paphuc — APPROVED** [review]

**Pull request merged**